packages/vite/src/node/plugins/resolve.ts

```diff
@@ -982,52 +982,15 @@ export function resolvePackageEntry(
       )
     }
 
-    // handle edge case with browser and module field semantics
-    if (!entryPoint && targetWeb && options.mainFields.includes('browser')) {
-      // check browser field
-      // https://github.com/defunctzombie/package-browser-field-spec
-      const browserEntry =
-        typeof data.browser === 'string'
-          ? data.browser
-          : isObject(data.browser) && data.browser['.']
-      if (browserEntry) {
-        // check if the package also has a "module" field.
-        if (
-          !options.isRequire &&
-          options.mainFields.includes('module') &&
-          typeof data.module === 'string' &&
-          data.module !== browserEntry
-        ) {
-          // if both are present, we may have a problem: some package points both
-          // to ESM, with "module" targeting Node.js, while some packages points
-          // "module" to browser ESM and "browser" to UMD/IIFE.
-          // the heuristics here is to actually read the browser entry when
-          // possible and check for hints of ESM. If it is not ESM, prefer "module"
-          // instead; Otherwise, assume it's ESM and use it.
-          const resolvedBrowserEntry = tryFsResolve(
-            path.join(dir, browserEntry),
-            options,
-          )
-          if (resolvedBrowserEntry) {
-            const content = fs.readFileSync(resolvedBrowserEntry, 'utf-8')
-            if (hasESMSyntax(content)) {
-              // likely ESM, prefer browser
-              entryPoint = browserEntry
-            } else {
-              // non-ESM, UMD or IIFE or CJS(!!! e.g. firebase 7.x), prefer module
-              entryPoint = data.module
-            }
-          }
-        } else {
-          entryPoint = browserEntry
-        }
-      }
-    }
-
     // fallback to mainFields if still not resolved
     if (!entryPoint) {
       for (const field of options.mainFields) {
-        if (field === 'browser') continue // already checked above
+        if (field === 'browser' && targetWeb) {
```

Comment:
I think the intention was to check `browser` and `module` without respecting their order. After this PR, if there is a custom `mainFields: ['module', 'browser']` with them inverted, before we were still resolving, and now we will directly use the `module` field. Maybe we should kick the edge case handling if we hit any of them. It should still fix your issue.

Something like having this out of the loop
```js
const mainFieldsIncludesBrowser = options.mainFields.includes('browser')
```

and then
```suggestion
        if (targetWeb && (field === 'browser' || (field === 'module' && mainFieldsIncludesBrowser) ) {
```

Reply:
this would be equally weird though, if the resolve order is "module" first, "browser" later, it should not prefer the browser value. Thats a configuration choice that should be respected. Tbh this whole thing probably belongs in a "weird-legacy-resolve" pre plugin that users can install as needed.

The logic inside only kicks in if browser has an entry. what if browser has no entry but module has one? then respecting the new order would work but keeping the old behavior would break. There is no univerally right approach with this. Ultimately i hope only very few libraries have pkg.browser and pkg.module and those two disagree :nauseated_face:

Reply:
If browser is before module, that should also be respected, even if it is the default. I agree with you, but we missed the opportunity to remove this edge case handling in vite 5, maybe we could in the next major (or try in a minor with a flag to add it back?). For getting this in a patch, probably better to keep it working as close as it was before. 

About the second paragraph. If there is no browser entry, the previous logic wont kick in, and the new either, no?

Reply:
This would only happen if package.json looked like this

```
"browser": {"foo":"dist/foo.js"},
"module":"dist/bar.js"
```

browser would not have an entry, but module does. If we went with the custom logic for module too, it would no longer resolve the valid entry from module.  

I really hope there are very few packages that do this kind of stuff and even less setups that put module before browser. Ultimately i'm fine with a patch that allows us to add the "svelte" condition first without this interfering. 

I do believe the more compatible thing to do is to follow the order in mainFields mostly and only massage the value of browser in place (and maybe module, but then module probably needs its own logic that works a bit differently than the one for browser).

Reply:
I let others chip in, maybe we should try to leave the edge case handling as you proposed.

Reply:
I didn't see the special browser handling this way initially, but I do see the point 🤔 I've gone back and forth while writing this comment. While I think patak's idea is great too, it does complicate the implementation.

So I'm leaning towards handling only if `browser` if found like this PR for now. The new `browser` string is only supported in Vite 5, and I doubt someone flips it today for any reason. If the reason is to have `module` take precedence over `browser`, that wouldn't work anyways before this PR.

Reply:
Ok, sounds good to me. I hope we can remove this special logic altogether at one point.